### PR TITLE
Loki: Improve query cache hits

### LINKED ISSUE #65994: Loki: Improve query caching
### Summary

Loki queries are cached based on the exact query expression. This PR lists some opportunities to improve caching.

### Solution

```[tasklist]
### Tasks
- [ ] trim query expressions before queries are executed
- [x] investigate if it would be worth sorting labels before queries are executed
- [x] investigate other opportunities to improve caching (e.g. sort parameters for `json` parser)
```

### Measurement

We already have metrics for cache hits and misses.

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
this takes the first steps towards our goal of trying to improve cache hits in loki

changes in this pr:

1. reorder selectors before running loki queries
2. trim queries of whitespace

> NOTE: THE QUERY DOES NOT CHANGE FOR THE USER, TO SEE THE REFLECTED CHANGE YOU WILL HAVE TO CHECK THE NETWORK TAB

```
input query:
{b_label="second", c_label="third", a_label="first"}

---

executed query:
{a_label="first", b_label="second", c_label="third"}
```

```
input query:
rate(  
    {b_label="second", a_label="first"}
  |=""
   |logfmt 
           |label=""
           [1s]        
) 

---

exectued query:
rate(
{a_label="first", b_label="second"}
|=""
|logfmt
|label=""
[1s]
)

```

**Which issue(s) does this PR fix?**:
Fixes #65994

### CODE REVIEW

File: public/app/plugins/datasource/loki/datasource.ts [hunk 1]
```diff
@@ -260,9 +261,18 @@ export class LokiDatasource
...
 
+    const transformed = queries.map((query) => {
+      const sorted = sortLabelSelectors(query);
+
+      const lines = sorted.expr.split('\n');
+      const trimmedLines = lines.map((line) => line.trim());
+      const trimmedExpr = trimmedLines.join('\n');
+      return { ...sorted, expr: trimmedExpr };
+    });
+
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.ts:272]: I would move all this to `queryUtils` and just call a function from there, so we don't keep piling up code in the datasource.

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 2]
```diff
@@ -293,3 +293,56 @@ export const getLokiQueryFromDataQuery = (query?: DataQuery): LokiQuery | undefi
...
+
+      if (labelNameA < labelNameB) {
+        return -1;
+      }
+
+      if (labelNameA > labelNameB) {
+        return 1;
+      }
+
+      return 0;
```
Comment (matyax) [inline public/app/plugins/datasource/loki/queryUtils.ts:325]: We could shorten this a little bit like:

```suggestion
      if (!labelNameA || !labelNameB) {
        return 0;
      }

      if (labelNameA < labelNameB) {
        return -1;
      }

      return (labelNameA > labelNameB) ? 1 : 0;
```

**Review by matyax — COMMENTED** [review]
Great work! both in the implementation and the PR title and description.
Left some comments and, as always, no test, no quest.

imagen

**Comment by gabor** [comment]
nice work 👍 . do you think there could be any complications for the user? (for example, Loki returns an error-message saying let' say error at position 10, but that position is now different because of reorder & trim?) i'm not saying such situation could happen, i just wonder.

**Comment by matyax** [comment]
> say error at position 10

Ouch, yes. That is a really good point. It would be extremely confusing.

**Comment by ivanahuckova** [comment]
I see in the linked issue that 
`investigate if it would be worth sorting labels before queries are executed` is checked out. I am wondering if we have any summary, because I can think of couple of places where this would complicate things for us:

- error messages coming from Loki with errors such as`error at position X` would be inaccurate
- in inspect query, you would have incorrect/different query than what you have written
- every time there is an issue/escalation related to querying, we would have to consider repositioned query

And I personally don't know, if there are many/significant amount of queries that have the same timestamps and labels and they are just reordered. I don't see this happening in dashboards at all, as the queries are saved. And in Explore I would expect that users have often different absolute timestamps. So I would like to see some data that this is actually an issue before we move on as I think this could really complicate things for debugging in the future.

**Review by ivanahuckova — CHANGES_REQUESTED** [review]
I see in the linked issue that `investigate if it would be worth sorting labels before queries are executed` is checked out. I am wondering if we have any summary, because I can think of couple of places where this would complicate things for us:

- error messages coming from Loki with errors such as `error at position X` would be inaccurate
- in inspect query, you would have incorrect/different query than what you have written
- every time there is an issue/escalation related to querying, we would have to consider repositioned query

And I personally don't know, if there are many/significant amount of queries that have the same timestamps and labels and they are just reordered. I don't see this happening in dashboards at all, as the queries are saved. And in Explore I would expect that users have often different absolute timestamps. So I would like to see some data that this is actually an issue before we move on as I think this could really complicate things for debugging in the future.

**Comment by gwdawson** [comment]
i spoke with @gabor about this before, the `investigate ...` was an oversight on my side thinking it wouldn't change much but as pointed out there are some pretty serious side affects to this.

i will put this on hold for now and we can discuss this more in mondays weekly.

**Pull request closed without merging**